kafka-client/src/main/java/dev/responsive/kafka/api/async/internals/contexts/AsyncContextRouter.java

```diff
@@ -0,0 +1,273 @@
+ * async thread's specific instance of the AsyncProcessorContext. Each execution thread
+ * in the async pool gets its own async context and the mapping is final and immutable
+ * once the async pool is initialized, allowing lock-free/contention-free routing and
+ * processor context access.
+ * <p>
+ * This context router allows us to work around the change in execution thread that
+ * occurs after an AsyncProcessor is initialized and processing begins (then flips
+ * again when the processor is closed).
+ * Since a processor context is only passed into a user's Processor implementation
+ * once, during #init, they will need to save a reference to the exact context
+ * object that's passed to #init, which is executed by the StreamThread. But after
+ * that point, it will be only AsyncThreads executing the user's #process method,
+ * and these async threads each have their individual context that is different
+ * from the one belonging to the StreamThread (and to each other's). To make sure
+ * any ProcessorContext calls the user makes inside #process reach the appropriate
+ * context belonging to the currently executing thread, instead of an actual context,
+ * we pass this "context router" in when invoking #init on the user's processor.
+ * This way, when they save a reference to the context, it is not the StreamThread's
+ * context but the router instead, and any calls made later in #process will be
+ * routed accordingly.
+ * <p>
+ * The StreamThread will flip the router to "processing mode" after it returns from
+ * the user's #init method. When the AsyncProcessor is eventually closed, the
+ * StreamThread will turn processing mode back off just before invoking the #close
+ * method of the user's processor.
+ * When in "processing mode", the router expects only AsyncThread access and should
+ * return only {@link AsyncThreadProcessorContext} instances.
+ * When processing mode is off, the router expects the executing thread to be the
+ * original StreamThread, and will only return the {@link StreamThreadProcessorContext}
+ * that belongs to it.
+ * <p>
+ * Threading notes:
+ * -Accessed by both the StreamThread and AsyncThreadPool, but only one or the other
+ *  depending on whether "processing mode" is on or off.
+ * -Although the async processing framework is responsible for setting up and managing
+ *  the state of this class, the context router is not itself for use by the framework
+ *  which should always have a handle on the specific context instance and pass things
+ *  off directly between if needed. In other words, this class is simply made to act
+ *  as an interface between the user and the underlying context objects. It is only
+ *  passed in to the user's #init and only serves to redirect the ProcessorContext
+ *  methods that the user invokes from their processor (whether in #init, #process,
+ *  or #close)
+ */
+@SuppressWarnings("checkstyle:overloadmethodsdeclarationorder")
+public class AsyncContextRouter<KOut, VOut>
```

Comment:
Suggestion on design - instead of doing our own routing, why not just use the existing `ThreadLocal` mechanism that java already provides. We can implement our own delegating context that delegates to a thread-local context set by the async processor and the async threads. They each set the thread-local context as appropriate before calling into the user's processor. You've essentially implemented thread-local yourself here, so this should cut the complexity down by using a native mechanism. I've prototyped what that could look like on this branch: https://github.com/responsivedev/responsive-pub/commit/a54e2f789e90312a070d4bf56554aa5727d68829#diff-650ebcfa1aec6e491543b24b2689d9f4b8708fe11bbe9710e633da3412cf2206

Reply:
On a related design note, I think handling of the context is one case that might justify tweaking the API that users need to use to one that is almost exactly the same as Processor API but that passes the context on every call to process. Tying the context to the thread calling process is a little brittle as it totally breaks if the user does any multi-threading in their processor. So it would require a slightly larger code-change from the user but its a bit safer. Not relevant for this PR though.

Reply:
Thanks for the ThreadLocal tip, I finished switching over from the manual router to the one you wrote that uses ThreadLocal. On the whole it's much cleaner although it does introduce the *slight* awkwardness when we create a new AsyncThreadProcessorContext for a specific AsyncThread when initializing a processor, since we can no longer immediately register the new context with the router class and have to instead wait for the AsyncThreead to notice and register it itself.. Anyways give the AsyncThread class another look, that's where all the new complexity is.

Reply:
Re: the possibility of multiple threads, I'm inclined to say that's not really supported to begin with so it's not a problem if it can't be done here. Or rather, none of the ProcessorContext actions are safe to perform from a different thread -- attempting to call #forward from another thread is definitely not safe.

I guess technically the metadata APIs are fine though, so that would be a valid use case that would break. But in that case the context is just acting as a metadata container, in which case users can just adapt their code to pass in the metadata (or a new simple container class) to the other threads instead of giving them a handle on the context itself.

Overall I feel like it would be weird to overly-optimize for what seems like a pretty rare case, especially if it means users who aren't using multiple threads will have more code changes just to save the users who are from a few code changes (eg wrapping metadata in their own container class). Does that make sense?